### PR TITLE
Be more strict classifying staggered intersections

### LINKED ISSUE #2824: Broken Testcase / Staggered Intersections might need revisiting
During my ongoing release of 5.3.3 I came across the following testcase:

https://github.com/Project-OSRM/osrm-backend/blob/da9d63a3845eced8c648a43ec6841611cb377274/features/testbot/via.feature#L266-L279

The cucumber scenario currently succeeds on master, but fails on 5.3.3. It should fail on both versions though, or is not helpful. What the test suggests is visible in the following picture:

vis-invalid

We query for the red path, which is obviously inferior to the green path. The green path is found (which would result in `ebbdcf, abcd, ebbdcf, ebbdcf`, but is collapsed again to `ebbdcf,ebbdcf` due to `https://github.com/Project-OSRM/osrm-backend/blob/da9d63a3845eced8c648a43ec6841611cb377274/src/engine/guidance/post_processing.cpp#L789` triggering here.

For the release of 5.3.3 I am marking this case as a todo. We should revisit how we want to handle this scenario and if the staggered intersection heuristic is valid. /cc @daniel-j-h

### PR DESCRIPTION
**Pull request by daniel-j-h**
# Issue

Staggered Intersections were triggering on data issues / uturn'ish and straight'ish "turns". No longer rely on direction modifier but instead use turn angles for being more strict in classification.

## Tasklist
 - [ ] review
 - [ ] adjust for for comments

## Requirements / Relations
- https://github.com/Project-OSRM/osrm-backend/issues/2824
- https://github.com/Project-OSRM/osrm-backend/pull/2825

### CODE REVIEW

File: features/testbot/via.feature [hunk 1]
```diff
@@ -264,9 +264,6 @@ Feature: Via points
             | 9,8,7     | cd,da,ab,bc,cd,cd,cd,da,ab,bc,cd,cd | 3000m +-1    |
 
     # See issue #2706
-    # this case is currently broken. It simply works as put here due to staggered intersections triggering a name collapse.
-    # See 2824 for further information
-    @todo
     Scenario: Incorrect ordering of nodes can produce multiple U-turns
```
Comment (MoKob) [inline features/testbot/via.feature]: right now this case is somewhat misleading. It only prints the way it does due to a hidden name change. I would say that the case needs to be adjusted to, in the end, print the correct path. This would require the instructions to have slight angles in them, though.
As the test is right now, it provides more confusion than anything else.

Comment (daniel-j-h) [inline features/testbot/via.feature]: Hm as this is unrelated to staggered intersections should I just leave the todo in then?

Comment (MoKob) [inline features/testbot/via.feature]: I guess keeping it as todo is the right thing to do for now

File: src/engine/guidance/post_processing.cpp [hunk 2]
```diff
@@ -528,12 +528,22 @@ bool isStaggeredIntersection(const RouteStep &previous, const RouteStep &current
...
+    // This more closely resembles what we understand as Staggered Intersection.
+    const auto is_right = [](const double angle) { return angle > 45 && angle < 135; };
+    const auto is_left = [](const double angle) { return angle > 225 && angle < 315; };
 
-    const auto left_right = isLeftTurn(previous.maneuver.instruction) && //
-                            isRightTurn(current.maneuver.instruction);
-    const auto right_left = isRightTurn(previous.maneuver.instruction) && //
-                            isLeftTurn(current.maneuver.instruction);
+    const auto left_right = is_left(angle(previous)) && is_right(angle(current));
+    const auto right_left = is_right(angle(previous)) && is_left(angle(current));
```
Comment (MoKob) [inline src/engine/guidance/post_processing.cpp]: should we also check for similar deviations or are combinations of sharp/slight still valid?
Have you looked at the combinations of extremes to see if they follow your idea of a staggered intersection?

Comment (daniel-j-h) [inline src/engine/guidance/post_processing.cpp]: The main use-case for this was the Staggered Intersection phenomena in the US where grid-like cities are the norm.

> Staggered intersections are very common in rural and exurban areas there. Often, as an area becomes built up, traffic engineers bend the roads to link them up directly, leaving a rump street.

Comment (daniel-j-h) [inline src/engine/guidance/post_processing.cpp]: See the following for an analysis in the US:
http://bl.ocks.org/anonymous/raw/5b44f131d86f3710d3b0e20b210c1798/

Comment (MoKob) [inline src/engine/guidance/post_processing.cpp]: We still should do an analysis of the non-obvious cases. If you, as was the case prior to this PR, allow 180 and 0 degree within the turns, strange situations might get detected.
What I am concerned about are cases like these:
sharp-staggered

If at some point we should increase the threshold (3 meters might be too small), we could end up combining very non-obvious turns.

I can't see from the above analysis if such cases would be a valid choice to classify them as staggered. We should at least see if these cases can exist, where they are and if they don't even exist make the angle more restricted. I know I said in chat `like 45 to 135 degreeish`, but I feel that we should test these bounds.

Comment (daniel-j-h) [inline src/engine/guidance/post_processing.cpp]: Yes the turn angles could also be steeper if we only target Staggered Intersections in the US.
